src/librustc_trans/glue.rs

```diff
@@ -397,60 +395,71 @@ pub fn size_and_align_of_dst<'a, 'tcx>(bcx: &BlockAndBuilder<'a, 'tcx>,
 }
 
 // Iterates through the elements of a structural type, dropping them.
-fn drop_structural_ty<'a, 'tcx>(cx: BlockAndBuilder<'a, 'tcx>,
-                                ptr: MaybeSizedValue,
-                                t: Ty<'tcx>)
-                                -> BlockAndBuilder<'a, 'tcx> {
-    fn iter_variant<'a, 'tcx>(cx: &BlockAndBuilder<'a, 'tcx>,
-                              t: Ty<'tcx>,
-                              av: adt::MaybeSizedValue,
-                              variant: &'tcx ty::VariantDef,
-                              substs: &Substs<'tcx>) {
+fn drop_structural_ty<'a, 'tcx>(cx: Builder<'a, 'tcx>, ptr: LvalueRef<'tcx>) -> Builder<'a, 'tcx> {
+    fn iter_variant<'a, 'tcx>(cx: &'a Builder<'a, 'tcx>,
+                              av: LvalueRef<'tcx>,
+                              adt_def: &'tcx AdtDef,
+                              variant_index: usize,
+                              substs: &'tcx Substs<'tcx>) {
+        let variant = &adt_def.variants[variant_index];
         let tcx = cx.tcx();
         for (i, field) in variant.fields.iter().enumerate() {
             let arg = monomorphize::field_ty(tcx, substs, field);
-            let field_ptr = adt::trans_field_ptr(&cx, t, av, Disr::from(variant.disr_val), i);
-            drop_ty(&cx, MaybeSizedValue::sized(field_ptr), arg);
+            let mut av = av.clone();
+            av.ty = LvalueTy::Downcast {
+                adt_def: adt_def,
+                substs: substs,
+                variant_index: variant_index,
+            };
```

Comment:
Why not do this in the caller? Or at least at the start of the function, only once.